Refactor product issues

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Part of https://github.com/woocommerce/google-listings-and-ads/issues/2146 & https://github.com/woocommerce/google-listings-and-ads/issues/2250

This PR fetches the product issues using the same background job used to fetch the product statuses. This allows us to reuse the WC products that were fetched while updating the product statuses.

The logic remains mostly unchanged from before. The main difference is that previously, Google IDs were stored in the database and used to fetch the issues. With the new sync mechanism, since we don't have the Google IDs until we fetch all product statuses, we now use the data from the product statuses to obtain the Google IDs and fetch the issues.

### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Set your PHP memory limit low so you can catch any heavy process. For example, setting this up to 128MB.
2. Call `GET wp-json/wc/gla/mc/product-statistics/refresh`
3. Go to WooCommerce -> Status -> Action Scheduler -> Search for `gla/jobs/update_merchant_product_statuses/process_item` see how the jobs get completed successfully.
4. See how the table `wp_gla_merchant_issues` gets populated.
5. Once all jobs are completed call `GET wp-json/wc/gla/mc/product-statistics` and see that you get the right stats.


### Additional details:
 - Account and pre-sync issues are not migrated yet.
 - After all the issues logic is refactored, I will remove any legacy code.
 - I wanted to address the scenario where the job fails and some issues are inserted. I used the "created at" field, but it appears that the field will be identical for all batches. Therefore, I'll need to see alternative solutions in a follow-up PR. Any suggestions are welcome too.
 
<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

>

jorgemd24 committed:
Add mc_id to the get_product_view_report response

jorgemd24 committed:
Refactor product issues

jorgemd24 committed:
Adjust merchant center report test

jorgemd24 committed:
Add delete_stale

jorgemd24 committed:
Refactor clear cache product statuses

jorgemd24 committed:
Add tests to refresh product issues

jorgemd24 committed:
Rename function and delete issues if the job fails so we don't have duplicate issues

jorgemd24 committed:
Fix issue with the comparision

jorgemd24 committed:
Add missing phpdoc

jorgemd24 committed:
Fix phpcs

jorgemd24 committed:
phpcs tests

jorgemd24 committed:
Add @var definition

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -235,14 +258,6 @@ public function maybe_refresh_status_data( bool $force_refresh = false ): void {
 		// Update account-level issues.
 		$this->refresh_account_issues();
 
-		// Update MC product issues and tabulate statistics in batches.
```
Seems like we are removing the product issues logic. Should we update also the description? Seems like we don't update stale status-related data regarding product issues anymore.

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -190,10 +191,32 @@ public function clear_cache(): void {
+	 *
+	 * @since x.x.x
+	 */
+	protected function delete_stale_issues( string $compare = '<' ) {
```
💅 Maybe we might add `: void `to be consistent with the rest of the class signatures.

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -27,6 +27,7 @@
 use Automattic\WooCommerce\GoogleListingsAndAds\Options\OptionsInterface;
 use DateTime;
 use Exception;
+use WC_Product;
```
💅  Unused `use Automattic\WooCommerce\GoogleListingsAndAds\Vendor\Google\Service\ShoppingContent\ProductStatus as GoogleProductStatus;`

puntope commented on src/DB/Table/MerchantIssueTable.php:
```diff
@@ -55,9 +55,10 @@ public static function get_raw_name(): string {
 	 * Delete stale issue records.
 	 *
 	 * @param DateTime $created_before Delete all records created before this.
+	 * @param string   $compare         Comparison operator to use. Default is '<'.
```
💅  Maybe we can delete the extra tabulation between $compare and the description.

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -354,67 +369,71 @@ protected function fetch_issues( string $type = null, int $per_page = 0, int $pa
+	 * Get MC product issues from a list of Product View statuses.
 	 *
-	 * @param string[] $google_ids
+	 * @param [] $statuses The list of Product View statuses.
```
Shouldn't be `array[]` instead of `[]`?

puntope replied:
Additionally. PHPDoc comment doesn't contain all the necessary @throws tags derivated from `$this->container->get`

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -354,67 +369,71 @@ protected function fetch_issues( string $type = null, int $per_page = 0, int $pa
-				'visibility' => get_post_meta( $wc_product_id, $visibility_meta_key ),
-				'parent_id'  => $wc_product->post_parent,
+			$product_issue_template = [
+				'product'              => html_entity_decode( $wc_product->get_name() ),
```
💅  Maybe `product_name` may be more accurate. (However was not caused in this PR)

jorgemd24 replied:
Yep, it would be more accurate but I'd prefer to keep it since it's tied to the column name in `wp_gla_merchant_issues`.

See: https://github.com/woocommerce/google-listings-and-ads/blob/7f3e1f57887658cadcc38f6795a36d1fb7fb1ccb/src/DB/Table/MerchantIssueTable.php#L33

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -488,52 +507,11 @@ protected function refresh_custom_merchant_issues() {
+	 * Refresh product issues in the merchant issues table.
 	 *
-	 * @param GoogleProductStatus[] $validated_mc_statuses Product statuses of validated products.
+	 * @param [] $product_issues Array of product issues.
```
💅  This might be `array` instead of `[]` also some throw tags missing.

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -618,31 +596,25 @@ protected function refresh_presync_product_issues(): void {
-	public function process_product_statuses( array $statuses ): void {
-		$product_repository = $this->container->get( ProductRepository::class );
-		$products           = $product_repository->find_by_ids_as_associative_array( array_column( $statuses, 'product_id' ) );
+	public function process_product_statuses( array $product_view_statuses ): void {
```
💅  Missing throw tags

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -668,13 +640,16 @@ public function process_product_statuses( array $statuses ): void {
+		$this->update_products_meta_with_mc_status();
+		$this->update_intermediate_product_statistics();
+
+		$product_issues = $this->get_product_issues( $product_view_statuses );
```
💅  WDYT about Saving $product_issues in the class we don't need to use params here?

jorgemd24 replied:
I'd prefer to keep the function independent of the instance state. This way, we can insert specific product issues without relying on the state.

puntope reviewed (commented):
Thanks @jorgemd24  for working on this. I tested following the indications inn the description and worked for me.

I left some comments regarding the code.

> I wanted to address the scenario where the job fails and some issues are inserted. I used the "created at" field, but it appears that the field will be identical for all batches. Therefore, I'll need to see alternative solutions in a follow-up PR. Any suggestions are welcome too.

Could you elaborate this? I'm not sure if I understand the issue.

jorgemd24 committed:
Update PHPDocs

jorgemd24 committed:
Update function description

jorgemd24 committed:
Revert delete_stale with compare query

jorgemd24 committed:
Add delete specific product issues

jorgemd24 committed:
Adjust tests for test_handle_failed_mc_statuses_fetching

jorgemd24 commented:
Thanks @puntope for the review! I addressed all your comments and replied to some of your suggestions. Regards:

> Could you elaborate this? I'm not sure if I understand the issue.

I was thinking a scenario where the job fails during inserting the issues. For instance, if the job inserts some issues but others fail, such as a query taking too long, the job itself fails, and Action Scheduler will retry it. This means it would reinsert the issues that were successfully added in the initial try. The `update_or_insert` will always insert the row because the key in the table is the `ID` and not the product_id and the issue so I decided to delete the issues for the products we were processing for that specific job, so if the job retries, it starts fresh without any stale data. See here: https://github.com/woocommerce/google-listings-and-ads/pull/2277/commits/c75eb4db0a2c9d3f4179696c25bf01ec1f354f5e

Could you please take another look at it? Thanks!

puntope commented:
> I decided to delete the issues for the products we were processing for that specific job, so if the job retries, it starts fresh without any stale data. 

Yes, I guess that might work. Other alternative option may be a duplicate cleanup after the AS Job batch finishes.

puntope commented on src/DB/Table/MerchantIssueTable.php:
```diff
@@ -61,6 +61,17 @@ public function delete_stale( DateTime $created_before ) {
+	 * @param array  $products_ids Array of product IDs to delete issues for.
+	 * @param string $source       The source of the issues. Default is 'mc'.
+	 */
+	public function delete_specific_product_issues( array $products_ids, string $source = 'mc' ) {
```
💅  We can add `:void `as a return type in the signature to be consistent with the other functions. `delete_stale`also missed it, but was not introduced in this PR tho

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -190,10 +191,30 @@ public function clear_cache(): void {
+	 *
+	 * @since x.x.x
+	 */
+	protected function delete_stale_issues() {
```
💅  Missed :void

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -190,10 +191,30 @@ public function clear_cache(): void {
+	 *
+	 * @since x.x.x
+	 */
+	public function clear_product_statuses_cache_and_issues(): void {
```
💅  Maybe we can add also some tests for this new logic.

puntope commented on src/MerchantCenter/MerchantStatuses.php:
```diff
@@ -354,67 +367,74 @@ protected function fetch_issues( string $type = null, int $per_page = 0, int $pa
 				continue;
 			}
+			// Unsynced issues shouldn't be shown.
+			if ( ChannelVisibility::DONT_SYNC_AND_SHOW === $wc_product->get_meta( $visibility_meta_key ) ) {
```
💅  Seems that we are not covering this case in tests.

puntope reviewed (approved):
Thanks @jorgemd24 for all the changes.

I approve this as there are no blockers anymore and it works as defined in the PR description.

I left tho some comments regarding code and test coverage.

jorgemd24 committed:
Add missing return types

jorgemd24 committed:
Add tests for clear_product_statuses_cache_and_issues

jorgemd24 committed:
Add tests for fetching issues for products with DONT_SYNC_AND_SHOW
